module/zfs/vdev_raidz.c

```diff
@@ -2744,6 +2916,140 @@ vdev_raidz_worst_error(raidz_row_t *rr)
+		m = (data[(n>>1) - 1] + data[n>>1]) >> 1;
+	else
+		m = data[((n + 1) >> 1) - 1];
+
+	return (m);
+}
+
+/*
+ * Calculate the outlier fence from a set of n latency values
+ *
+ * fence = Q3 + 2 x (Q3 - Q1)
+ */
+static uint64_t
+latency_quartiles_fence(const uint64_t *data, size_t n)
+{
+	uint64_t q1, q3;
+
+	q1 = latency_median_value(&data[0], n>>1);
+	if (n % 2 == 0)
+		q3 = latency_median_value(&data[n>>1], n>>1);
+	else
+		q3 = latency_median_value(&data[(n+1) >> 1], n>>1);
+
+	uint64_t iqr = q3 - q1;
+	uint64_t fence = q3 + iqr;
+
+	return (fence);
+}
+
+static void
+latency_sort(uint64_t *samples, int size) {
+	/* Insertion sort but size is small */
+	for (int i = 1; i < size; i++) {
+		uint64_t val = samples[i];
+		int j = i;
+		while (j > 0 && samples[j - 1] > val) {
+			samples[j] = samples[j - 1];
+			j--;
+		}
+		samples[j] = val;
+	}
+}
+
+#define	STACK_SAMPLES	32
+#define	MIN_LAT_SAMPLES	4
+
+/*
+ * Check for any latency outlier from latest set of child reads.
+ *
+ * Uses a Tukey's fence, with K = 2, for detecting extreme outliers. This
+ * rule defines extreme outliers as data points outside the fence of the
+ * third quartile plus two times the Interquartile Range (IQR). This range
+ * is the distance between the first and third quartile.
+ */
+noinline static void
+vdev_child_slow_outlier(zio_t *zio)
+{
+	vdev_t *vd = zio->io_vd;
+	if (raid_read_sit_out_secs == 0 || vd->vdev_children < MIN_LAT_SAMPLES)
+		return;
+
+	spa_t *spa = zio->io_spa;
+	if (spa_load_state(spa) == SPA_LOAD_TRYIMPORT ||
+	    spa_load_state(spa) == SPA_LOAD_RECOVER ||
+	    (spa_load_state(spa) != SPA_LOAD_NONE &&
+	    spa->spa_last_open_failed)) {
+		return;
+	}
+
+	hrtime_t now = gethrtime();
+	if ((now - vd->vdev_last_latency_check) <
+	    MSEC2NSEC(raid_outlier_check_interval_ms)) {
+		return;
+	}
+
+	vd->vdev_last_latency_check = now;
+
+	int samples = vd->vdev_children;
+	uint64_t data[STACK_SAMPLES];
+	uint64_t *lat_data;
+
+	if (samples > STACK_SAMPLES)
+		lat_data = kmem_alloc(sizeof (uint64_t) * samples, KM_SLEEP);
+	else
+		lat_data = &data[0];
+
+	uint64_t max = 0;
+	vdev_t *svd = NULL; /* suspect vdev */
+	for (int c = 0; c < samples; c++) {
+		vdev_t *cvd = vd->vdev_child[c];
+
+		if (cvd->vdev_read_sit_out_expire > 0) {
+			atomic_store_64(&cvd->vdev_recent_latency, 0);
+			goto out;
+		}
+
+		lat_data[c] = atomic_load_64(&cvd->vdev_recent_latency);
+
+		/* wait until all disks have been read from */
+		if (lat_data[c] == 0)
+			goto out;
+
+		/* keep track of the vdev with largest value */
+		if (lat_data[c] > max) {
+			max = lat_data[c];
+			svd = cvd;
+		}
+	}
+
+	latency_sort(lat_data, samples);
+	uint64_t fence = latency_quartiles_fence(lat_data, samples);
+	if (lat_data[samples - 1] > fence) {
+		/*
+		 * Keep track of how many times this child has had
+		 * an outlier read. A disk that persitently has a
+		 * higer than peers outlier count will be considered
+		 * a slow disk.
+		 */
+		atomic_add_64(&svd->vdev_outlier_count, 1);
```

Comment:
With small number of children and with only one random sample from each I really doubt this math can be statistically meaningful.